GcExtensions.md

```diff
@@ -0,0 +1,353 @@
+
+The mark function argument (`markfunc`) and the finalizer function
+(`finalizefunc`) take function pointer arguments; the finalizer function
+can be `NULL`.
+
+The `haspointers` parameter should be non-zero if
+the object may contain references to Julia objects; the `large` parameter
+should be non-zero if the size is greater than the value returned by
+`jl_gc_max_internal_obj_size()` and zero otherwise. If the object can be
+both larger or not, then two distinct foreign types need to be created, one
+for the case where the size is less than or equal and one for the case where
+it is larger than the value of `jl_gc_max_internal_obj_size()`.
+
+```
+typedef uintptr_t (*jl_markfunc_t)(jl_ptls_t ptls, jl_value_t *obj);
+typedef void (*jl_finalizefunc_t)(jl_value_t *obj);
+jl_datatype_t *jl_new_foreign_type(
+  jl_sym_t *name,
+  jl_module_t *module,
+  jl_datatype_t *super,
+  jl_markfunc_t markfunc,
+  jl_finalizefunc_t finalizefunc,
+  int haspointers,
+  int large
+);
+size_t jl_gc_max_internal_obj_size(void);
+```
+
+*Performance impact:* Custom mark functions need to be called during the
+performance-critical mark loop of the garbage collector. In order to
+avoid overhead for the other cases, the code is engineered to consider
+such objects as the last possible option in the existing if-else chains.
+To accomplish that, such foreign types use the existing
+`jl_datatype_layout_t` structure, with `fielddesc_type` set to `3`,
+which is looked at after the other data types and the other alternatives
+for `fielddesc_type`.
+
+The mark function gets passed a pointer to thread-local storage (`ptls`)
+and the object to be marked (which will be of the type defined through
+`jl_new_foreign_type()`. The `ptls` argument is an optimization so that
+`jl_get_ptls_states()` does not need to be called unnecessarily during
+the mark loop.
+
+The mark function implementation also uses `jl_gc_mark_queue_obj()` to
+mark objects, as with the root scanner; however, in contrast to marking
+roots, the return value cannot be ignored. Per object, the mark function
+should count how often `jl_gc_mark_queue_obj()` for subjects return
+non-zero values and return that number. If an object has no subobjects,
+the mark function should return zero.
+
+This information is relevant for the generational part of garbage
+collection. The return value of `jl_gc_mark_queue_obj()` is non-zero
+if a young generation object has been marked. When the mark function
+has been called for an old object and the mark function returns a
+non-zero value (thus showing how many young objects have been marked
+from the old one one), the GC knows to update its internal data
+structures accordingly.
+
+For an example of this, see the `gcext` test in the Julia repository,
+which defines a couple of such custom mark functions.
+
+To enable C finalization for such an object, the function
+`jl_gc_set_needs_foreign_finalizer()` has to be called for an
```

Comment:
Why is this not implied/automatic?

Reply:
What do you mean with that? The system must somehow be able to tell if finalizers should be run for a given object -- or perhaps rather the converse: we want to avoid the overhead for objects that don't need a finalizer.

I am probably missing something, could you perhaps elaborate? Thanks

Reply:
The reason is that one may not want to finalize all objects of a given foreign type.

Reply:
K, sure, if there's usecase for that. Otherwise I think it's less error prone to just define a special function to allocate foreign object with the finalizer automatically added.

Reply:
Unfortunately, this can change over the lifetime of an object -- i.e. an object may be created w/o needing a finalizer, but then later on it turns out that it needs one.

Just like it can happen with other Julia objects, via `jl_gc_add_finalizer`... And in fact, note that this is currently the implementation of `jl_gc_set_needs_foreign_finalizer`:
```C
JL_DLLEXPORT void jl_gc_set_needs_foreign_finalizer(jl_value_t *obj)
 {
     jl_gc_add_finalizer((jl_value_t *)obj, jl_nothing);
 }
 ```